libraries/lib-wave-track/Sequence.cpp

```diff
@@ -1377,6 +1393,12 @@ bool Sequence::Append(
    return result;
 }
 
+void Sequence::InitPlaybackCache() {
+   mPlaybackCache = std::make_unique<SequenceSampleCache>(
+      std::bind(&Sequence::DoGet, this, _1, _2, _3, _4, _5), mNumSamples,
+      mMaxSamples);
+}
```

Comment:
As we discussed, a question is whether the cache is invalidated in sufficient places.

The cache depends on mNumSamples and mMaxSamples, so it should be invalidated at least when either of those changes.  (Though also when contents change).

I haven't studied the later commits, but even this one is incomplete by that criterion.

 `mMaxSamples` can change in `ConvertToSampleFormat`.  (Also `HandleXMLTag`, but that is deserialization of the structure from disk, so in fact just one initialization of the cache at the end of `HandleXMLEndTag`  is sufficient.)

`mNumSamples` may change where `AppendBlock` is called -- surprise!, the third argument is a reference.  Also in `Paste`, `Delete`.  Also `CommitChangesIfConsistent`, `AppendBlocksIfConsistent` but you covered those.